File: android/app/src/main/java/com/videoconverter/android/lan/LanLibrary.kt
package com.videoconverter.android.lan

import com.videoconverter.android.domain.Job
import com.videoconverter.android.domain.JobStatus
import com.videoconverter.android.domain.resolveConfig
import com.videoconverter.android.ui.historyDateLabel
import java.time.ZoneId

enum class LanLibraryTab {
    Video, Audio, Image, Document
}

fun LanLibraryTab.wireName(): String = when (this) {
    LanLibraryTab.Video -> "video"
    LanLibraryTab.Audio -> "audio"
    LanLibraryTab.Image -> "image"
    LanLibraryTab.Document -> "document"
}

fun lanLibraryTabFor(kind: LanPreviewKind): LanLibraryTab = when (kind) {
    LanPreviewKind.Video -> LanLibraryTab.Video
    LanPreviewKind.Audio -> LanLibraryTab.Audio
    LanPreviewKind.Image -> LanLibraryTab.Image
    LanPreviewKind.Pdf, LanPreviewKind.File -> LanLibraryTab.Document
}

data class LanLibraryItem(
    val jobId: String,
    val index: Int,
    val path: String,
    val kind: LanPreviewKind,
    val tab: LanLibraryTab,
    val label: String,
    val format: String,
    val detail: String,
    val needsIndex: Boolean,
)

fun lanLibraryItems(
    jobs: List<Job>,
    zone: ZoneId = ZoneId.systemDefault(),
    untitled: String = "Untitled",
    fileExists: (String) -> Boolean,
): List<LanLibraryItem> {
    val out = ArrayList<LanLibraryItem>()
    for (job in jobs.asReversed()) {
        if (job.status != JobStatus.Completed) continue
        val paths = jobOutputPaths(job)
        for ((index, path) in paths.withIndex()) {
            if (path.isBlank() || !fileExists(path)) continue
            val label = lanPreviewFileName(path, job, untitled)
            val kind = lanPreviewKind(label)
            val format = resolveConfig(job.config).getOrNull()?.container
                ?: label.substringAfterLast('.', job.config.preset)
            val detail = listOfNotNull(
                format.takeIf { it.isNotBlank() },
                historyDateLabel(job.createdAtEpochMs, zone),
            ).joinToString(" · ")
            out += LanLibraryItem(
                jobId = job.id,
                index = index,
                path = path,
                kind = kind,
                tab = lanLibraryTabFor(kind),
                label = label,
                format = format,
                detail = detail,
                needsIndex = paths.size > 1 || index > 0,
            )
        }
    }
    return out
}

fun lanLibraryItemsFor(items: List<LanLibraryItem>, tab: LanLibraryTab): List<LanLibraryItem> =
    items.filter { it.tab == tab }

fun lanDefaultLibraryTab(items: List<LanLibraryItem>): LanLibraryTab {
    val order = listOf(
        LanLibraryTab.Video,
        LanLibraryTab.Audio,
        LanLibraryTab.Image,
        LanLibraryTab.Document,
    )
    return order.firstOrNull { tab -> items.any { it.tab == tab } } ?: LanLibraryTab.Video
}

File: ios/LiteTrans/Domain/LanLibrary.swift
import Foundation

public enum LanLibraryTab: String, CaseIterable, Sendable {
    case video, audio, image, document
}

public func lanLibraryTabWireName(_ tab: LanLibraryTab) -> String {
    tab.rawValue
}

public func lanLibraryTabFor(_ kind: LanPreviewKind) -> LanLibraryTab {
    switch kind {
    case .video: return .video
    case .audio: return .audio
    case .image: return .image
    case .pdf, .file: return .document
    }
}

public struct LanLibraryItem: Equatable, Sendable {
    public var jobId: String
    public var index: Int
    public var path: String
    public var kind: LanPreviewKind
    public var tab: LanLibraryTab
    public var label: String
    public var format: String
    public var needsIndex: Bool

    public init(
        jobId: String,
        index: Int,
        path: String,
        kind: LanPreviewKind,
        tab: LanLibraryTab,
        label: String,
        format: String,
        needsIndex: Bool
    ) {
        self.jobId = jobId
        self.index = index
        self.path = path
        self.kind = kind
        self.tab = tab
        self.label = label
        self.format = format
        self.needsIndex = needsIndex
    }
}

public func lanLibraryItems(_ jobs: [Job], fileExists: (String) -> Bool) -> [LanLibraryItem] {
    var out: [LanLibraryItem] = []
    for job in jobs.reversed() {
        guard job.status == .completed else { continue }
        let paths = jobOutputPaths(job)
        for (index, path) in paths.enumerated() {
            if path.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty || !fileExists(path) { continue }
            let label = lanPreviewFileName(path, job: job)
            let kind = lanPreviewKind(label)
            let format = (try? resolveConfig(job.config).container) ?? lanFileExtension(label).nilIfEmpty ?? job.config.preset
            out.append(
                LanLibraryItem(
                    jobId: job.id,
                    index: index,
                    path: path,
                    kind: kind,
                    tab: lanLibraryTabFor(kind),
                    label: label,
                    format: format,
                    needsIndex: paths.count > 1 || index > 0
                )
            )
        }
    }
    return out
}

public func lanLibraryItemsFor(_ items: [LanLibraryItem], tab: LanLibraryTab) -> [LanLibraryItem] {
    items.filter { $0.tab == tab }
}

public func lanDefaultLibraryTab(_ items: [LanLibraryItem]) -> LanLibraryTab {
    let order: [LanLibraryTab] = [.video, .audio, .image, .document]
    return order.first { tab in items.contains { $0.tab == tab } } ?? .video
}

public func lanLibraryTabLabel(_ tab: LanLibraryTab, copy: LanHistoryCopy) -> String {
    switch tab {
    case .video: return copy.video
    case .audio: return copy.audio
    case .image: return copy.image
    case .document: return copy.document
    }
}

public func lanLibraryEmptyLabel(_ tab: LanLibraryTab, copy: LanHistoryCopy) -> String {
    switch tab {
    case .video: return copy.emptyVideo
    case .audio: return copy.emptyAudio
    case .image: return copy.emptyImage
    case .document: return copy.emptyDocument
    }
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}
